service_container/injection_types.rst

```diff
@@ -100,6 +100,103 @@ working with optional dependencies. It is also more difficult to use in
+Another possible injection in to use a method which return ``static``,
+this approach allow you to make a service immutable::
+
+    // ...
+
+    class NewsletterManager
+    {
+        private $mailer;
+
+        /**
+         * @required
+         * @return static
+         */
+        public function withMailer(MailerInterface $mailer)
+        {
+            $new = clone $this;
+            $new->mailer = $mailer;
+
+            return $new;
+        }
+
+        // ...
+    }
+
+In order to use the full potential of this approach, you can define a third argument
+which allow the container to return the newly created service:
+
+.. configuration-block::
+
+    .. code-block:: yaml
+
+        # config/services.yaml
+       services:
+            # ...
+
+            app.newsletter_manager:
+                class: App\Mail\NewsletterManager
+                calls:
+                    - [withMailer, ['@mailer'], true]
+
+    .. code-block:: xml
+
+        <!-- config/services.xml -->
+        <?xml version="1.0" encoding="UTF-8" ?>
+        <container xmlns="http://symfony.com/schema/dic/services"
+            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+            xsi:schemaLocation="http://symfony.com/schema/dic/services
+                http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+            <services>
+                <!-- ... -->
+
+                <service id="app.newsletter_manager" class="App\Mail\NewsletterManager">
+                    <call method="withMailer" use-result="true">
+                        <argument type="service" id="mailer" />
+                    </call>
+                </service>
+            </services>
+        </container>
+
+    .. code-block:: php
+
+        // config/services.php
+        use App\Mail\NewsletterManager;
+        use Symfony\Component\DependencyInjection\Reference;
+
+        // ...
+        $container->register('app.newsletter_manager', NewsletterManager::class)
+            ->addMethodCall('withMailer', [new Reference('mailer')], true);
+
+.. note::
+
+    This type of injection requires that you add a ``@return static`` docblock in order
+    for the container to be capable of registering the method.
+
+This approach is useful if you need to configure your service according to your needs,
+so, what are the advantages?
+
+* Your service becomes immutable, as the container will return a new object,
+  the initial service stays clean and unchanged.
+
+* You can easily change the injected service as long as it respect the interface/type
+  asked by the initial service.
+
+The disadvantages are:
+
+* As the ``@return static`` docblock is required by the container to
+  understand that the method return a new object,
+  you can found that adding docblock for a single method isn't adapted or
+  link your code to the container.
+
+* As this approach force the container to create a new object
+  once the method is called, you can found hard to debug and test your code.
```

Comment:
Not sure to agree. Why do you think so?

Reply:
For me, using a newly cloned service can lead to harder debug time for after thinking about it, my suggestions seems stupid, that's not a change on the core logic of the service, just a clone of the original service with a new dependency, my bad :/

Reply:
Removed :)